### DISCUSSION TITLE
Clipping textures to exact coordinates possible?

### DISCUSSION THREAD

**Discussion by denisreiter** [citygml4j/citygml-tools#45] [Q&A] [answered]
Hello,

First of all thank you for providing this amazing toolset!
I'm currently trying to apply the function clip-textures to the 3D model of Berlin (https://www.businesslocationcenter.de/downloadportal): 

I want to clip the textures of the model to the exact wall coordinates so that I have a perfectly aligned front view of a wall, roof etc. like they are visualized when loading the model in any viewer (1st image is an example).
![Screenshot 2023-07-12 143103](https://github.com/citygml4j/citygml-tools/assets/64582999/e502641d-1cae-4dc8-ae6d-6e0c20153f92)

The textures that are bundled with the 3D model are distorded images, and running the clip-textures function just returns the the same images again (2nd image example), with no clipping done at all.
![tex_2591910](https://github.com/citygml4j/citygml-tools/assets/64582999/ccc3b6aa-4791-41a9-ae6e-5477adbcf196)

Is it possible to do that with this toolset? If I understand the schema of CityGML correctly, there should be texture coordinates associated with every image, so that they can be at least displayed correctly. From my understanding, the clip-textures function should exactly do what I want, but it seems I did something wrong?

Thanks in advance!

**Comment by clausnagel** [comment, marked as answer]
Hi @denisreiter, as far as I understand, you want to rectify your oblique images. Unfortunately, the `clip-textures` command of citygml-tools will not help you with this task. This command is rather meant for scenarios where a large texture image is assigned to a surface but only a small portion of the texture space is actually used for texturing this surface. Based on the texture coordinates in the CityGML file, the `clip-textures` command identifies that part of the texture image that is really needed for texturing the surface and cuts off the rest. Thus, the texture images are reduced in size but no other image operation is applied. 

If no clipping is applied to a texture image (like in your example), this means all pixels are needed for texturing the surface.
